@jkeenan I dont think @haarg's patch will change this behavior.

The error message you are encountering is from some crappy panic-style code I did to automatically extract change file entries from CHANGES and other change files. I tested it against all the distributions that we have in perl, which is why I did not bother improving the error message, I assumed it would never happen with the modules we have. Bad me. So I will follow up with a better patch for this message so it is more self-explanatory what is going wrong, and a way to defang the exception from the command line so it doesn't break a flow like the one that you have created. After all it is preventing you from updating the module because it cant parse the Changes file which is not designed to be parsed anyway, so really it shouldnt be a fatal exception. I manually edited sync-with-cpan so it outputs the following:

```
panic: No changes in change file Acme-Homer/Changes for this module?
(This indicates an issue with sync-with-cpan)
old_version: 0.01
new_version: 0.000002
```

Now, for the actual issue you are seeing, what is happening is that sync-with-cpan uses the CPAN generated file 02packages.details.txt for version lookups (this is actually the reason why sometimes you need to wait a day to upgrade a module that has been uploaded to CPAN, the file is updated daily only). You can find this file in your /tmp directory after running the script.

This file contains version data in "normalized floating point form" (i dont know a better way to express this), so the dist you wanted to update is versioned as v0.0.2, BUT cpan normalizes that to 0.000002, you can see this here:

```
$ grep Acme::Homer /tmp/02packages.details.txt 
Acme::Homer                    0.000002  D/DM/DMUEY/Acme-Homer-v0.0.2.tar.gz
```

sync-with-cpan uses the version number in the second column in that line. It does NOT use the version number from the file. The logic for extracting the changes entry looks for something that looks like that value. If we check the changes file we see this:

```
$ cat cpan/Acme-Homer/Changes 
Revision history for Perl extension Acme::Homer

0.0.1  Sat Dec 17 19:17:27 2005
	- original version; created by h2xs 1.22 with options
		-AXc -n Acme::Homer

0.0.2  Tue Dec 20 10:13:23 2005
	- POD fixes and addition of Flanders section
```

As you can see the file does not contain the normalized 0.000002 value, but rather contains an entry for `0.0.2 ....`. Interestingly  it also contains an entry for 0.0.1 but actually THAT version was 0.01, so in fact this modules module version migration has broken the versioning model, as 0.000002 < 0.01. So sync-with-cpan would not find version 0.01 in the changes file EITHER.

*SO*, does @haarg's patch fix sync-with-cpan with regard to v0.0.x style versions, as far as I can tell not completely, and in fact there are issues with the intrinsic structure of the CPAN data files that sync-with-cpan uses to handle this type of version, second, the test case you chose unfortunately is broken as well.

Ill take a look at improving some of this, but given that CPAN itself is part of the problem here we have to think carefully about how this works. 

As far as I can tell @haarg's patch is fine, it just  isnt a complete solution to the problem at hand.

--- a/Porting/core-cpan-diff
+++ b/Porting/core-cpan-diff
@@ -435,9 +435,8 @@ sub relatively_mapped {
 
 sub distro_base {
     my $d = shift;
-    $d =~ s/\.tar\.gz$//;
-    $d =~ s/\.gip$//;
-    $d =~ s/[\d\-_\.]+$//;
+    my $tail_pat = qr/\.(?:tar\.(?:g?z|bz2|Z)|zip|tgz|tbz)/;
+    $d =~ s{-v?([0-9._]+(?:-TRIAL[0-9]*)?)$tail_pat\z}{};
     return $d;
 }
 

--- a/Porting/exec-bit.txt
+++ b/Porting/exec-bit.txt
@@ -51,6 +51,7 @@ Porting/checkcfgvar.pl
 Porting/checkpodencoding.pl
 Porting/cmpVERSION.pl
 Porting/config_h.pl
+Porting/core-cpan-diff
 Porting/corecpan.pl
 Porting/corelist-perldelta.pl
 Porting/corelist.pl

--- a/Porting/sync-with-cpan
+++ b/Porting/sync-with-cpan
@@ -410,9 +410,10 @@ chdir "$type_dir";
 my  $pkg_dir      = $files[0];
     $pkg_dir      =~ s!.*/!!;
 
-my $tail_pat = qr/(?:\.tar\.gz|\.tgz|\.zip)\z/;
+my $tail_pat = qr/\.(?:tar\.(?:g?z|bz2|Z)|zip|tgz|tbz)/;
+my $version_pat = qr/-v?([0-9._]+(?:-TRIAL[0-9]*)?)$tail_pat\z/;
 
-my ($old_version) = $distribution =~ /-([0-9._]+(?:-TRIAL[0-9]*)?)$tail_pat/;
+my ($old_version) = $distribution =~ $version_pat;
 
 if (!$old_version) {
     die "WTF: failed to parse old version from '$distribution'\n";
@@ -457,7 +458,7 @@ if (defined $tarball) {
     die "Tarball $tarball does not exist\n" if !-e $tarball;
     die "Tarball $tarball is not a plain file\n" if !-f _;
     $new_file     = $tarball;
-    $new_version  = $version // ($new_file =~ /-([0-9._]+(?:-TRIAL[0-9]*)?)$tail_pat/) [0];
+    $new_version  = $version // ($new_file =~ $version_pat) [0];
     die "Blead and that tarball both have version $new_version of $module\n"
         if $new_version eq $old_version;
 }
@@ -504,7 +505,7 @@ system git => 'clean', '-dfxq', $pkg_dir;
 say "Unpacking $new_file";
 Archive::Tar->extract_archive( $new_file );
 
-(my $new_dir = basename($new_file)) =~ s/$tail_pat//;
+(my $new_dir = basename($new_file)) =~ s/$tail_pat\z//;
 # ensure 'make' will update all files
 my $t= time;
 for my $file (find_type_f($new_dir)) {

